docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+1.  Validation for debug disallows some `Container` fields such as `resources`
+1.  If successful, client is responsible for `attach` to container if
+    appropriate
+
+#### Changes to External Tools
+
+Tools that examine pod state by inspecting the pod spec must be updated to also
+inspect `PodStatus.DebugContainerStatuses`.
+
+### Mounting Pod Volumes
+
+When debugging an application it's frequently useful to be able to inspect
+volumes shared within a pod. We can support an option to mount all pod volumes
+by name to a user-configurable mount point defaulting to `/mnt/volumes`. For
+*Copy Debug* this can be done when generating the new pod spec. The kubelet's
+debug handler would perform a similar action for *Running Debug*.
+
+### Cross Mounting Container Images
+
+Depending on resolution of [#831](https://issues.k8s.io/831), the final step
+will be to implement cross-container filesystem image mounts. Implementing that
+feature would almost certainly require amending the Container Runtime Interface
+to support mounting the filesystem of a different container. Once implemented it
+should be easy to amend the debug handler to mount the other container
+filesystems by name in (e.g.) `/mnt/containers`.
+
+## User Stories
+
+### Debugging
+
+Samantha has a service that consists of a statically compiled Go binary running
+in a minimal container that is suddenly having trouble connecting to an internal
+service. Her troubleshooting session might resemble:
+
+```
+% kubectl get pods
+NAME          READY     STATUS    RESTARTS   AGE
+neato-5thn0   1/1       Running   0          1d
+% kubectl debug -it -m debian -p neato-5thn0 -- bash
+root@debug-image-neato-5thn0:/# cat /etc/resolv.conf
+search default.svc.cluster.local svc.cluster.local cluster.local
+nameserver 10.155.240.10
+options ndots:5
+root@debug-image-neato-5thn0:/# dig @10.155.240.10 neato.svc.cluster.local.
+
+; <<>> DiG 9.9.5-9+deb8u6-Debian <<>> @10.155.240.10 neato.svc.cluster.local.
+; (1 server found)
+;; global options: +cmd
+;; connection timed out; no servers could be reached
+```
+
+This leads Samantha to discover that the cluster's DNS service isn't responding.
+
+### Automation
+
+Abe is a security engineer tasked with running security audits across all of his
+company's running containers. Even though his company has no standard base
+image, he's able to audit all containers using:
+
+```
+% for pod in $(kubectl get -o name pod); do
+    kubectl debug -m gcr.io/neato/security-audit -p $pod /security-audit.sh
+  done
+```
+
+## Alternatives Considered
+
+### Mutable Pod Spec
+
+Rather than adding an operation to have Kubernetes attach a pod we could instead
+make the pod spec mutable so the client can generate an update adding a
+container. `SyncPod()` has no issues adding the container to the pod at that
+point, but an immutable pod spec has been a basic assumption in Kubernetes thus
+far and changing it carries risk. It's preferable to keep the pod spec immutable
```

Comment:
This will change with `vertical pod auto-scaling` for example. Before we preclude pod spec updates, can we at-least list the issues with mutating pod spec?